### PR TITLE
Gradient descent

### PR DESCRIPTION
**Pull request by ckolbPTB**
I created a notebook which does gradient descent for MR, PET and CT. Somewhere towards the end of the notebook you find the parameter `curr_modality` which can be set to either `mr`, `pet` or `ct` in order to use the different modalities. 

The step-size `tau` has to be adapted for each modality. Values which I found to work ok, where 0.03 for `mr`, 0.3 for `pet` and 0.1 for `ct`. Would be great if somebody could try it.

### CODE REVIEW

**Review by evgueni-ovtchinnikov — CHANGES_REQUESTED** [review]
Some suggestions made for moving from `as_array` + `fill` + `numpy` algebra to SIRF algebra.

**Comment by KrisThielemans** [comment]
> Some suggestions made for moving from `as_array` + `fill` + `numpy` algebra to SIRF algebra.

great. I guess it was based on older examples/notebooks? Maybe the PET notebook still does that (even though the SIRF example doesn't)

File: notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb [hunk 1]
```diff
@@ -0,0 +1,791 @@
...
+    "    # Perform gradient ascent step\n",
+    "    curr_image_estimate_array =  curr_image_estimate_array + step_size * grad_array\n",
+    "\n",
+    "    # In PET we have to ensure values are positive. \n",
+    "    # We check if data is not complex and therefore PET or CT\n",
+    "    if not curr_image_estimate_array.dtype in [numpy.complex, numpy.complex64]:\n",
+    "        curr_image_estimate_array = make_positive(curr_image_estimate_array)\n",
+    "    \n",
+    "    # Update the image\n",
+    "    curr_image_estimate.fill(curr_image_estimate_array)\n",
```
[the remark this answers is not in the document]
Comment (KrisThielemans) [reply notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb:727]: The `maximum` trick is a bit hard to understand. What about putting that in `make_positive` function, such that this code is cleaner?

This is therefore actually constrained gradient descent. I believe the CT guys don't do that (although they could). @epapoutsellis , @paskino?

The comment text seems inconsistent at the moment. I guess it would have to see "In PET or CT, we have to ensure that..."

File: notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb [hunk 2]
```diff
@@ -0,0 +1,791 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Plot a slice for different iterations\n",
+    "centre_slice = image_iteration.shape[1]//2\n",
+    "plt.figure();\n",
+    "for i in range(4):\n",
+    "    curr_it = i*8\n",
+    "    if len(image_iteration.shape) == 4: # PET, MR\n",
+    "        plot_2d_image([2,2,i+1], image_iteration[curr_it,centre_slice,:,:], 'It '+str(curr_it), cmap=\"viridis\")\n",
```
[the remark this answers is not in the document]
Comment (KrisThielemans) [reply notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb:760]: needs `abs`? (I didn't check)

[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb:760]: see previous comment where `image` is defined

**Comment by ckolbPTB** [comment]
@evgueni-ovtchinnikov Thank you very much for the suggestions to make it more object orientated rather than using numpy arrays. I made all the ammendments except for the check if the data is complex. Here I had to leave the old check on the numpy array, because the CT data does not have the method is_complex().

**Comment by KrisThielemans** [comment]
great. I haven't checked the updated version yet, but what about making this a bit less "automatic" and let people do it by hand

```
if not image_data.as_array().dtype in [numpy.complex, numpy.complex64]:
```
->
```
if is_MRI:
```
and have a cell above it setting `is_MRI` and tell them what to set it.

I think it's less cryptic for an introductory notebook.

**Comment by ckolbPTB** [comment]
I added a check based on the methodology which is select by the user.

**Comment by KrisThielemans** [comment]
ready to merge? (we can fine-tune later if need be, but this way it's easier for people to see what it here)

**Review by KrisThielemans — APPROVED** [review]

File: notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb [hunk 3]
```diff
@@ -0,0 +1,789 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Author: Christoph Kolbitsch, Edoardo Pasca\n",
+    "First version: 23rd of April 2021  \n",
+    "\n",
+    "CCP PETMR Synergistic Image Reconstruction Framework (SIRF).  \n",
+    "Copyright 2015 - 2017 Rutherford Appleton Laboratory STFC.  \n",
```
Comment (ashgillman) [inline notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb:25]:
```suggestion
    "Copyright 2021 Rutherford Appleton Laboratory STFC.  \n",
```

File: notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb [hunk 4]
```diff
@@ -0,0 +1,789 @@
...
+    "Copyright 2015 - 2019 University College London.   \n",
```
Comment (ashgillman) [inline notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb:26]:
```suggestion
```

**Review by ashgillman — COMMENTED** [review]
Looks good, I wonder if I can replicate with the PyTorch GD

File: notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb (hunk 3 above)
Comment (KrisThielemans) [reply notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb:25]: I've kept these. This file contains copies of code that was written ages ago, so they are appropriate.

File: notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb (hunk 4 above)
Comment (KrisThielemans) [reply notebooks/Synergistic/gradient_descent_mr_pet_ct.ipynb:26]: I've kept these. This file contains copies of code that was written ages ago, so they are appropriate.
